# Assignment 2 Naïve Bayes and Logistic Regression
Q1. Using the Iris dataset in Jupyter Notebook,
1.	Extract X as all columns except the last column and Y as last column. (1 mark)

2.	Create a Naïve Bayes Classification model to predict the species of flowers with explanation of parameters. (2 marks)

3.	Compute the accuracy and confusion matrix. (1 marks)


# Import the Librarires 

In [27]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
import matplotlib.pyplot as plt

# Load the Scikit-learn dataset to a pandas Dataframe :

In [28]:
data = load_iris()
iris_data = pd.DataFrame(data['data'], columns=data['feature_names'])
iris_data['target'] = data['target']
iris_data.head()  # Inspecting dataset

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [29]:
# Inspecting datatypes of attributes in our dataset 
iris_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
target               150 non-null int32
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [30]:
# Inspecting shape of our dataset
iris_data.shape

(150, 5)

In [31]:
#No missing values present
iris_data.isna().sum()

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
dtype: int64

# Assiging values to X and y co-ordinates :

In [32]:
X = iris_data.iloc[:,1:4]
X

sepal width (cm)  petal length (cm)  petal width (cm)
0                 3.5                1.4               0.2
1                 3.0                1.4               0.2
2                 3.2                1.3               0.2
3                 3.1                1.5               0.2
4                 3.6                1.4               0.2
..                ...                ...               ...
145               3.0                5.2               2.3
146               2.5                5.0               1.9
147               3.0                5.2               2.0
148               3.4                5.4               2.3
149               3.0                5.1               1.8

[150 rows x 3 columns]

In [33]:
y = iris_data[['target']]
y

target
0         0
1         0
2         0
3         0
4         0
..      ...
145       2
146       2
147       2
148       2
149       2

[150 rows x 1 columns]

In [34]:
y.shape

(150, 1)

In [35]:
import matplotlib.colors as colors
import seaborn as sns
%matplotlib inline
sns.set()

In [36]:
# Splitting the dataset into the Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [37]:
# As we need a normal distribution let us use StandardScaler to scale our features to a normal distribution
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [38]:
# Fitting Naive Bayes to the Training set
classifier = GaussianNB()
classifier.fit(X_train, y_train.values.ravel())

GaussianNB(priors=None, var_smoothing=1e-09)

In [39]:
# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Confusion Matrix and Accuracy 

In [48]:
cm = confusion_matrix(y_test, y_pred)
print('confusion matrix :\n',cm)

confusion matrix :
 [[13  0  0]
 [ 0 15  1]
 [ 0  0  9]]


In [46]:
acc = accuracy_score(y_test, y_pred)
print('accuracy :',acc)

accuracy : 0.9736842105263158


In [45]:
# Finally verifying our Classification report
cr = classification_report(y_test, y_pred)
print(cr)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       1.00      0.94      0.97        16
           2       0.90      1.00      0.95         9

    accuracy                           0.97        38
   macro avg       0.97      0.98      0.97        38
weighted avg       0.98      0.97      0.97        38



We can there by conclude that our model has 97% accuray in identifying the type of Iris flower by using its sepal 
and petal lenght and width. Thereby classification report provides same accuracy and precision scores thereby making it a good model